### DISCUSSION TITLE
UI state management

### DISCUSSION THREAD

**Discussion by ViMaSter** [Studio-Lovelies/GG-JointJustice-Unity#195] [Ideas]
Opening this thread so I'm not derailing the conversation in #170 any further and we have this a place for general questions and potential use-cases for it.

---

A few weeks ago I proposed using [a global JSON object](https://github.com/ViMaSter/unity-databinding) for UI management across the game.  
I could see this kind of system coming in handy for some things like:

#### Passing data across scenes
This could be useful for things like chapter selection or when we want to migrate the current court record when switching between scenes.

#### Decoupling "scripts that render data" and "scripts that modify data
Imagine a script which just "draws a button based on whatever JSON-data we have" and another which "changes the JSON-data if we hover over it / navigate to it using the keyboard".  
Both scripts no longer have to be aware of each other and can also be tested independently.

**Comment by ViMaSter** [comment]
Continuing from https://github.com/Studio-Lovelies/GG-JointJustice-Unity/discussions/170#discussioncomment-2177521:

> In regard to the JSON databinding, I don't exactly understand the advantage of that as opposed to using static fields in C#. Seems like they both allow you to access data from anywhere, but one is quite complex and I don't understand it.

[That is actually literally what is happening under the hood.](https://github.com/ViMaSter/unity-databinding/blob/main/Packages/com.vimaster.databinding/Scripts/Document.cs#L126) The only difference being that it's not static itself, so every element that wants to access it still needs a direct reference to the `Document` (so you could have multiple if you ever needed to).

> I've been meaning to take a look at the databinding package you linked to try and understand what it's doing, but heh, time and stuff :p

The easiest way to visualize it, is having a `.json`-file that exists for the entirety of the game and anything can write to and read from it and can also be notified if a specific value changes.

> That said, if it persists data between opening and closing the game that could be very useful for audio settings etc,

I think the SaveFile functionality we have is the better place for that. That way there's only one place for "things persisted across sessions".

> but the scene to load isn't really something that needs to persist between opening and closing the game

I definitely agree with you on that! The data-binding JSON wouldn't persist between opening and closing the game, but it wouldn't be cleared until the game is closed or if a script clears a specific key.

**Reply by PhoebeMitchell** [reply]
How exactly does the data persist between scenes? I created a little test script

```
private Document document;
    
    private IEnumerator Start()
    {
        document = GetComponent<Document>();
        try
        {
            Debug.Log(document.Get<string>("test"));
        }
        catch (Exception)
        {
            Debug.Log("Error");
        }
        document.Set("test", "test");

        yield return new WaitForSeconds(1);
        SceneManager.LoadScene(SceneManager.GetActiveScene().name);
    }
```

In theory it should print "Error" once, then "test" repeatedly. But it just prints out error. Am I misunderstanding? This aligns with my understanding of the code; you create a new JObject for every instance of Document, so how would it persist?
<img width="250" alt="Screenshot 2022-02-17 at 07 44 11" src="https://user-images.githubusercontent.com/67964179/154428955-7a78ae9a-3e5f-4774-b1bd-b26da9fd1b5e.png">

I agree the SaveFile stuff would be better for saving :p "Why not use that?" would have been my next question.

I massively dislike string dependency, _especially_ multiple string dependencies for one thing (e.g, you reference UIElement multiple times in components). There is no easy way to update all of them, there's no error highlighting, and it could cause errors that are difficult to fix, for example if we update all paths but forget one. I would much rather have references to objects through code and the Unity editor.

All in all I feel this is a very complex solution to a problem that doesn't really exist, and I think the current way of dealing with things is more intuitive.

**Reply by ViMaSter** [reply]
> How exactly does the data persist between scenes? I created a little test script

That's a miscommunication on my part. : D  
The behavior you're experiencing is definitely correct: If the GameObject dies, the `Document` (like all other Components) dies with it.  
My idea would be to have a GameObject that isn't destroyed on load, to keep the `Document` alive and let that be **the only thing** that ever persists across scenes. (I think that keeping the amount of GameObjects and Components that persist across scenes low, makes it easy to follow the flow of the game through the code.)   
However it's nothing that is forced by the `Document` directly.

> I massively dislike string dependency, especially multiple string dependencies for one thing (e.g, you reference UIElement multiple times in Components). There is no easy way to update all of them, there's no error highlighting, and it could cause errors that are difficult to fix, for example if we update all paths but forget one. I would much rather have references to objects through code and the Unity editor.

I definitely can see that and I agree with the sentiment in general.  
As a possible option, we could have a `public const string` and use references to that instead; that way the actual key is never used directly and could easily be changed (like we do for the [PlayerPrefs key](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/blob/develop/unity-ggjj/Assets/Scripts/SaveFiles/PlayerPrefsProxy.cs#L11)).  
Serializing and deserializing things from a pretty type-safe language like C# to and from JSON (which is fairly loose in comparison), does introduce a layer of abstraction though, so I can see where you're coming from.  
It's probably purely a personal preference coming from web dev, where your only option for "persistent storage" of any form is a JSON-blob in Local storage or Cookies; and even session state management tools for React/Vue/Angular effectively also only have a "reactive JSON blob" in the background.

> All in all I feel this is a very complex solution to a problem that doesn't really exist, and I think the current way of dealing with things is more intuitive.

That's totally fair! You're also the one working on most (if not all) UI stuff at the moment, so I'm not going to push back on it further, if you don't feel like there's a need for it. : D  

Just to make the advantages I see a bit more clear, these are two specific "problems" I see at the moment this kind of system could solve:
- The ability to easily test renderers and logic separately:
  Since the only data source for any sort of renderer is the `Document`, a render would only ever use the `Document` or Unity-internal UI Components. (We could literally hard code a JSON object in tests and then assert something like an `EvidenceMenuRenderer` to setup `Image` and `Text` Components accordingly.)
  Similarly, for the logic-side of things, these classes wouldn't even require anything Unity-specific; only a `Document` to write to and whatever logic they use to modify the data inside the `Document` (which is favorable, as `EditMode` tests require no scene and tend to run faster).

  and

- Instead of managing multiple GameObjects with different Components that might need to be marked as `DontDestroyOnLoad` to persist data across scenes (like the `NarrativeScriptPlaylist` and Chapter Select-screen), we could use the JSON blob as our only "storage that persists across scenes".  
  Like I mentioned at the beginning of the post, I think having as few GameObjects and Components persist as possible, reduces the amount of "things to keep in mind", when a scene starts up.

That said, I hear you loud and clear! If you're comfortable with working with UI at the moment, this very well may be a solution to a problem we don't actually have and premature optimization is the root of all evil. : D

**Reply by PhoebeMitchell** [reply]
Thanks for the detailed reply. I worry I may be being unreasonably resistant to this idea. If other members of the team have input I'd happily hear them out.

How would the Document be referenced in a new scene? The only things I think of are Find methods or a singleton. I was under the impression that an advantage of this method would be that we can avoid those things. Currently I struggle to understand how this would be advantageous as opposed to having a class to store data marked DontDestroyOnLoad that is also accessed via singleton or Find.

**Reply by ViMaSter** [reply]
> I worry I may be being unreasonably resistant to this idea.

Absolutely not! If you feel there's issues or I'm overengineering things, please always voice those concerns! : D  
With pretty much all things I can see where you're coming from, so it's good you bring those things up, since I failed to address them initially.

> How would the Document be referenced in a new scene? The only things I think of are Find methods or a singleton. I was under the impression that an advantage of this method would be that we can avoid those things.

Not really, I'm afraid. It's just a way to store data; how we manage its lifetime and access is open for us to decide.  
The main advantage is that it's JSON; it would remove the need to update that singleton if we want to introduce new fields or modify how existing fields are used.

Say we add a countdown timer:  
With the `Document` being JSON, a logic component could simply write to `countdown.time_remaining` and the renderer could read from that path.  
If we were to build this with a static instance of a C# class containing that data, we'd always have to manually add a field or property to it before other components can use it. *  
As that's written in C# code, it also invites more complex logic to be added, to what should only be a plain data object. (As `struct`s can also contain methods or properties, we can't limit it from a language perspective.)
```cs
// root level...
public TimeSpan timeRemaining;

// ...or if we wanted to store it nested...
public class CountdownData
{
    public TimeSpan timeRemaining;
}
public CountdownData Countdown = new CountdownData();

// nothing aside from a human review also prevents us from doing something like this
public string FormattedTimeRemaining { 
    get => string.Format("{0:D2}:{1:D2}", duration.Minutes, duration.Seconds)
} 
```

<sup>* = That said, under the string dependency lens you mentioned, this might actually be considered a good thing, since we're now referencing fields. I personally would find adding the code above for each field or nested group of fields to be a bit cumbersome, but I can definitely see it both ways.
